### PR TITLE
Feature/live start at

### PR DESCRIPTION
**Pull request by pauloSantosMobileTW**
## Goal
Allow DVR Videos to be initiated in a position before the live edge

## How to test
1. Run tests

### CODE REVIEW

File: Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (paulogama14) [inline Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme:1]: Maybe we can avoid adding this file or is there a special reason for this?

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 2]
```diff
@@ -264,9 +268,19 @@ open class AVFoundationPlayback: Playback {
...
     private func seekToStartAtIfNeeded() {
-        if canStartAt {
-            seek(startAt)
-        }
+        guard canStartAt else { return }
+        
+        seek(startAt)
+    }
+    
+    private func seekToLiveStartTimeIfNeeded() {
```
Comment (paulogama14) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:276]: I think we are adding more than one responsibility in this method. An indication of that is the `ifNeeded` in the name.

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 3]
```diff
@@ -264,9 +268,19 @@ open class AVFoundationPlayback: Playback {
...
+        guard canLiveStartAt, let liveStartTime = liveStartTime else { return }
+        
+        let positionToSeek = liveStartTime - epochDvrWindowStart
+        
+        if positionToSeek < 0 || positionToSeek > duration { return }
```
Comment (paulogama14) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:281]: I think this rule can be encapsulated in a method or var to be more expressive. What do you think?

File: Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+<Scheme
```
Comment (helbertgs) [inline Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme:2]: I think you don't need to merge this file

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 3 above)
Comment (helbertgs) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:281]: I have the same opinion

File: Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme (hunk 1 above)
Comment (dezouzart) [reply Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme:1]: this change is about showing test target on xcode menu. IMHO it helps. But I can revert it if you want to.

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 2 above)
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: Done

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 3 above)
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: Done

File: Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme (hunk 4 above)
Comment (pauloSantosMobileTW) [reply Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme:2]: See comment above

File: Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme (hunk 1 above)
Comment (paulogama14) [reply Clappr.xcodeproj/xcshareddata/xcschemes/Clappr_Tests.xcscheme:1]: Okey-Dokey.
=]

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 5]
```diff
@@ -227,6 +227,10 @@ open class AVFoundationPlayback: Playback {
     private var canStartAt: Bool {
         return startAt != 0.0 && playbackType == .vod
     }
+    
+    private var canLiveStartTime: Bool {
+        return isEpochInsideDVRWindow(liveStartTime) && isDvrAvailable
```
Comment (BruunoMM) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:232]: We could short circuit this if we invert the clause's order. Avoiding doing the math if DVR isn't available.

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 6]
```diff
@@ -264,9 +268,17 @@ open class AVFoundationPlayback: Playback {
...
-        if canStartAt {
-            seek(startAt)
-        }
+        guard canStartAt else { return }
+        
+        seek(startAt)
+    }
+    
+    private func seekToLiveStartTime() {
+        guard let liveStartTime = liveStartTime else { return }
```
Comment (BruunoMM) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:277]: There's no need to do this check here. You already check that in `canLiveStartTime`

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift [hunk 7]
```diff
@@ -227,6 +227,10 @@ open class AVFoundationPlayback: Playback {
     private var canStartAt: Bool {
         return startAt != 0.0 && playbackType == .vod
     }
+    
+    private var canLiveStartTime: Bool {
```
Comment (BruunoMM) [inline Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:231]: What do you think about giving a more descriptive name to this variable? I think the word `time` here does not fit. Something around `canStartLiveAt`, `canLiveStartAtTime` or even `canLiveStartAtGivenTime`.

File: Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackTests.swift [hunk 8]
```diff
@@ -780,6 +780,127 @@ class AVFoundationPlaybackTests: QuickSpec {
...
+                                    
+                                    playback.on(Event.didUpdatePosition.rawValue) { userInfo in
+                                        didCallUpdatePosition = true
+                                        liveStartTime = userInfo!["position"] as! Double
+                                    }
+                                    
+                                    playback.handleDvrAvailabilityChange()
+                                    
+                                    expect(didCallUpdatePosition).toEventually(beTrue())
+                                    expect(liveStartTime).toEventually(equal(500))
```
Comment (BruunoMM) [inline Tests/Clappr_Tests/Classes/Plugin/Playback/AVFoundationPlaybackTests.swift:820]: Hey guys, I found a bit hard to understand why the result would be 500. Do you think is there a more descriptive way to write this test?

My suggestion would be to pass a `currentDate` to `setupLiveDvr`, so the math to calculate the result (expected value) would be contained inside the test itself. WDYT?

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 6 above)
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift:277]: Swift does not allow make a calculation with a Double? and a Double, this guard statement is required here

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 5 above)
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: Done

File: Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift (hunk 7 above)
Comment (pauloSantosMobileTW) [reply Sources/Clappr/Classes/Plugin/Playback/AVFoundationPlayback.swift]: Done

**Pull request merged**